### PR TITLE
Issue90 generate retroprov nanopubs

### PR DESCRIPTION
**Pull request by raar1**
Addresses both #90 and #151

Can test this using the noodles_fw.ipynb example notebook. Tests have been added, but currently the multithreaded execution seems to hang indefinitely when run in pytest (otherwise runs fine in notebook or as regular script). Will need to investigate this further, but in separate issue so we don't block creation of the image processing pipelines etc.

### CODE REVIEW

File: fairworkflows/fairworkflow.py [hunk 1]
```diff
@@ -395,11 +404,36 @@ def execute(self, num_threads=1):
...
+        # execution down a bit. In future we will either write our own version of
+        # the runner, to grab that extra prov without any db writing, or we will
+        # read it from the database and convert it to RDF then.
+        with TemporaryDirectory() as td:
+            temp_db_fname = Path(td) / 'retro_prov.db'
+            result = run_parallel(self.noodles_promise, n_threads=num_threads,
+                                  registry=registry, db_file=temp_db_fname,
+                                  always_cache=True, echo_log=False)
+
+        # Generate the retrospective provenance as a (nano-) Publication object
```
Comment (svenvanderburg) [inline fairworkflows/fairworkflow.py:424]: I think it would be cleaner if we factor out a method for this.

File: tests/test_fairworkflow.py [hunk 2]
```diff
@@ -332,3 +332,68 @@ def test_publish_as_nanopub_no_modifications(self, mock_publish, test_workflow):
...
+
+        fw = my_workflow(1, 4, 3)
+        assert isinstance(fw, FairWorkflow)
+
+        # TODO: Find out why execution hangs only when running in pytest. May be to do with the threading.
+        #result, prov = fw.execute(num_threads=1)
+        #assert result == -66
+        #assert isinstance(prov, Publication)
+
+    def test_workflow_serialization(self):
```
Comment (svenvanderburg) [inline tests/test_fairworkflow.py:379]: I guess you test serialization of a **complex** object here, maybe make that explicit.

**Review by svenvanderburg — APPROVED** [review]
Nice changes! I like using the nanopub object for now to wrap the retrospective prov. In (the perhaps non-existing) future we could think about having a `RetroProv` object that inherits from `RdfWrapper`. It might be confusing for users that there is a different API to the Retro prov than the fair step and fair workflow. Additionally, potentially people might want to edit the retro prov or publish it to something else than nanopub. What do you think?

File: fairworkflows/fairworkflow.py (hunk 1 above)
Comment (raar1) [reply fairworkflows/fairworkflow.py:424]: Good idea. Done

File: tests/test_fairworkflow.py (hunk 2 above)
Comment (raar1) [reply tests/test_fairworkflow.py]: Agreed. Changed the name.

**Comment by raar1** [comment]
> Nice changes! I like using the nanopub object for now to wrap the retrospective prov. In (the perhaps non-existing) future we could think about having a `RetroProv` object that inherits from `RdfWrapper`. It might be confusing for users that there is a different API to the Retro prov than the fair step and fair workflow. Additionally, potentially people might want to edit the retro prov or publish it to something else than nanopub. What do you think?

I agree completely. In fact, I was caught out by this myself, thinking that there would be a ```publish_as_nanopub()``` method. For now I will leave it like this until some of the more pressing issues are dealt with, but then yes we should give it its own object. Shouldn't be too difficult.

**Pull request merged**